### PR TITLE
Minpack test_hybrd: Fix ASR Pass issues

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Fixes https://github.com/lfortran/lfortran/issues/1362
Fixes https://github.com/lfortran/lfortran/issues/1636
Fixes https://github.com/lfortran/lfortran/issues/1564

### CODE REVIEW

File: examples/expr2.f90 [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
-end program
+    subroutine compare_solutions()
+        real, dimension(:), allocatable :: temp
+        temp = solution()
+        print *, temp
+    end subroutine compare_solutions
+
+    pure function solution() result(x)
+        real,dimension(:),allocatable :: x
+        x = [0.10,0.10,1., .1, .23]
```
Comment (Thirumalai-Shaktivel) [inline examples/expr2.f90:16]: Here, I'm using the `ArrayConstant` size to allocate memory.

File: examples/expr2.f90 [hunk 2]
```diff
@@ -1,9 +1,31 @@
...
+    call compare_solutions()
 
-x = (2+3)*5
-print *, x
+    contains
 
-end program
+    subroutine compare_solutions()
+        real, dimension(:), allocatable :: temp
+        temp = solution()
```
Comment (Thirumalai-Shaktivel) [inline examples/expr2.f90:10]: Use the UBound of the `__libasr__created__var__0__func_call_res`, to allocate memory for `temp`.
This seems to fix the allocatable seg_fault.

I'm not sure if this is a clean solution.

**Comment by Pranavchiku** [comment]
Thank you Thirumalai, for taking this up! Looks like #1631 will also get fixed with this PR, it seems same as #1564.

**Comment by czgdp1807** [comment]
The `minpack` CI fails because `test_hybrd.f90` segfaults. Reason is because of the following,

```python
__libasr__created__section__0_slice: (Variable 2 __libasr__created__section__0_slice [n] Local () () Default (Real 8 [((IntegerConstant 1 (Integer 4 [])) (IntegerBinOp (IntegerBinOp (IntegerBinOp (Cast (Var 2 n) IntegerToInteger (Integer 4 []) ()) Sub (Cast (IntegerConstant 1 (Integer 4 [])) IntegerToInteger (Integer 4 []) ()) (Integer 4 []) ()) Div (Cast (IntegerConstant 1 (Integer 4 [])) IntegerToInteger (Integer 4 []) ()) (Integer 4 []) ()) Add (IntegerConstant 1 (Integer 4 [])) (Integer 4 []) ()))]) () Source Public Required .false.),
```

If you notice the size of `__libasr__created__section__0_slice` depends upon `n` (which is not a `parameter` a.k.a constant). This leads to the following LLVM code,

```llvm
%n = alloca i32, align 4
  %__libasr__created__section__0_slice = alloca %array.1, align 8
  %2 = load i32, i32* %n, align 4
  %3 = sub i32 %2, 1
  %4 = udiv i32 %3, 1
  %5 = add i32 %4, 1
  %6 = getelementptr %array.1, %array.1* %__libasr__created__section__0_slice, i32 0, i32 1
  store i32 0, i32* %6, align 4
  %7 = getelementptr %array.1, %array.1* %__libasr__created__section__0_slice, i32 0, i32 2
  %8 = alloca i32, align 4
  store i32 1, i32* %8, align 4
  %9 = load i32, i32* %8, align 4
  %10 = alloca %dimension_descriptor, i32 %9, align 8
  %11 = getelementptr %array.1, %array.1* %__libasr__created__section__0_slice, i32 0, i32 4
  store i32 1, i32* %11, align 4
  store %dimension_descriptor* %10, %dimension_descriptor** %7, align 8
  %12 = load %dimension_descriptor*, %dimension_descriptor** %7, align 8
  %13 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %12, i32 0
  %14 = getelementptr %dimension_descriptor, %dimension_descriptor* %13, i32 0, i32 0
  %15 = getelementptr %dimension_descriptor, %dimension_descriptor* %13, i32 0, i32 1
  %16 = getelementptr %dimension_descriptor, %dimension_descriptor* %13, i32 0, i32 2
  store i32 1, i32* %14, align 4
  store i32 1, i32* %15, align 4
  store i32 %5, i32* %16, align 4
  %17 = alloca i32, align 4
  %18 = mul i32 1, %5
  store i32 %18, i32* %17, align 4
  %19 = getelementptr %array.1, %array.1* %__libasr__created__section__0_slice, i32 0, i32 0
  %20 = load i32, i32* %17, align 4
  %21 = alloca double, i32 %20, align 8
  store double* %21, double** %19, align 8
```

Value of `%20` depends upon `%n` which just contains a garbage value. So we are just allocating a random amount of memory for `__libasr__created__section__0_slice`. How to fix this?

1. ASR verify should make sure that not a single local `Variable` node which has a size depending on another local `Variable` node is present. This will enforce that we don't get segmentation due to allocation of garbage amount of memory.
2. `__libasr__created__section__0_slice` is a generated variable in `asr_slice` pass. The size of such generated variables can be expressed in the form of ASR expressions. However its not always the case that such ASR expressions evaluate to compile time constants. In such situations the above ASR verify suggested check would fail (which is necessary). To avoid that we need to make `__libasr__created__section__0_slice` an `allocatable` and add `allocate` statements to reserve the right amount of memory for such generated variables. Another way is to make the type of  `__libasr__created__section__0_slice` same as the array which is being sliced (this is wasteful because sliced portion always consumes lesser memory than the array being sliced). So making an `allocatable` is a better option in my opinion (if we ignore ownership issues associated with allocatables).

@certik Let's discuss this in our meeting today.

**Comment by certik** [comment]
The ultimate optimized ASR has to look like this, nothing radical about it. The only question is if it is easier to do it right away, or in steps. Whatever is easier to maintain.

**Comment by czgdp1807** [comment]
> The only question is if it is easier to do it right away, or in steps.

That's the radical part about it. We have only one (and the best) option is to do it right away (because once we update the ASR verify then `arr_slice.cpp` will have to be removed). I will begin it on Friday this week (Wednesday and Thursday I have to work on #1604).

**Comment by certik** [comment]
Perfect.

**Comment by czgdp1807** [comment]
So here's the scene. The solution I mentioned in https://github.com/lfortran/lfortran/pull/1630#issuecomment-1531589612 is not complete because we need to support the following case as well,

```fortran
program array_slice_01
implicit none

integer, target :: a(5, 5, 5), sum_slice
a = 20
sum_slice = sum_array_slice(a(:, 2, :))
print *, sum_slice
print *, a

contains

function sum_array_slice(array) result(r)
integer, intent(inout) :: array(:, :)
integer :: i, j
integer :: r
r = 0
do i = lbound(array, 1), ubound(array, 1)
    do j = lbound(array, 2), ubound(array, 2)
      r = r + array(i, j)
      array(i, j) = 19
    end do
end do
end function

end program
```

**Output**

```zsh
(arm-compilers) 15:16:06:~/lfortran_project/lfortran % ./a.out                                             
         500
          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20
```

You can see the original array is also modified even though we passed the slice of original array to it. How do we deal with this? By transforming the above fortran code into the following,

```fortran
program array_slice_01
implicit none

integer, target :: a(5, 5, 5), sum_slice
integer, pointer :: b(:, :)
b => a(:, 2, :)
a = 20
sum_slice = sum_array_slice(b)
print *, sum_slice
print *, a

contains

function sum_array_slice(array) result(r)
integer, intent(inout) :: array(:, :)
integer :: i, j
integer :: r
r = 0
do i = lbound(array, 1), ubound(array, 1)
    do j = lbound(array, 2), ubound(array, 2)
      r = r + array(i, j)
      array(i, j) = 19
    end do
end do
end function

end program
```

**Output**

```zsh
(arm-compilers) 18:29:35:~/lfortran_project/lfortran % ./a.out
         500
          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20          19          19          19          19          19          20          20          20          20          20          20          20          20          20          20          20          20          20          20          20
```

So we have to support the second case compulsorily because `gfortran` compiles it. Now if we have to do it anyways then we can support all kinds of slicing by transforming the following,

```fortran
c = a(:, 2, 4:10) + b
```

into,

```fortran
integer, pointer :: a_slice(:, :)
a_slice => a(:, 2, 4:10) ! Add target attribute to `a`
c = a_slice + b
```

So that's it, no need to create allocatables. And it is valid fortran code as can be seen from above.

**Comment by czgdp1807** [comment]
In the backend we will be required to change the we generate code for array indexing. Right now we don't use offset in descriptors, we ignore them. But I think to support `a_slice => a(:, 2, 4:10)`, we will have to set the offset in descriptor of `a_slice` otherwise `a_slice` would index the data of `a` incorrectly. We also need to store the base (i.e., location of first element of `a_slice` in the data of `a`). By offset hear I mean the number of elements to be skipped to jump onto the next element in the same dimension.

So assume that `integer :: a(5, 6, 10)`. Now, offset of each dimension of `a` (assuming column major order),

| 1 | 2 | 3  |
|---|---|----|
| 1 | 6 | 60 |

So when indexed `a(2, 3, 5)`, using CMO, the location of the element in data of `a` will be `(5 - 1) * 60 + (3 - 1) * 6 + 1 * (2 - 1)` i.e., `253`. Base i.e., location of first element will be `0`.

Now for `a_slice`, computing offsets is tricky. Here's what I did.

1. Offset of dimension 1 of `a_slice`: First element of dimension 1 is `a(1, 2, 4)` and second element of dimension 1 is `a(2, 2, 4)`. So, taking the difference of `Location(a(2, 2, 4)) - Location(a(1, 2, 4))` we get, `187 - 186` i.e., `1`. So offset of dimension 1 of `a_slice` comes out to be `1`.
2. Offset of dimension 2 of `a_slice`: First element of dimension 2 is `a(1, 2, 4)` and second element of dimension 2 is `a(1, 2, 5)`. So again taking the difference of `Location(a(1, 2, 5)) - Location(a(1, 2, 4))` we get, `246 - 186` i.e., `60`.
3. Base of `a_slice` is `Location(a(1, 2, 4))` i.e., `186`.

Now let's index `a_slice(3, 7)` (equivalent to `a(3, 2, 7)`). `Location(a_slice(3, 7))` is `186 + (7 - 4) * 60 + (3 - 1) * 1` we get, `368`. Now doing `Location(a(3, 2, 7))`, we get `(7 - 1) * 60 + (2 - 1) * 6 + (3 - 1) * 1` i.e., `368`. Within the same block of data we get `Location(a_slice(3, 7)) == Location(a(3, 2, 7))` as exactly same. So I think the algorithm for indexing any array becomes,

`(i0 - l0) * offset0 + (i1 - l1) * offset1 + ... + (in - ln) * offsetn`. (i0, i1, ..., in) are the indices, (l0, l1, ... ln) are the lower bounds of each dimension and (offset0, offset1, ... offsetn)  is the offset of each dimension.

Now, for doing `a_slice => a(:, 2, 4:10)`, just compute the offset of only the sliced dimensions and set it to the dimension of `a_slice`. The base would be the `Location(a(l0, 2, 4))`. In general the first element of the sliced array.

Please let me know if the above makes sense to you @certik. If you think the above logic is correct then I would go ahead and update LLVM backend to use the above indexing algorithms and then support the cases in https://github.com/lfortran/lfortran/pull/1630#issuecomment-1536231630 and then finally update `arr_slice.cpp` to use the approach I described.

**Comment by certik** [comment]
Yes, using the pointer trick to "reindex" I think is fine, we need to support it anyway.

Indeed, if an array uses a descriptor, I think we have to support the "general" indexing, as you described.

I wonder if it is time to promote the array descriptor into the type system, since that seems to be a clean way to do this: https://github.com/lfortran/lfortran/issues/1601.

**Comment by czgdp1807** [comment]
> Yes, using the pointer trick to "reindex" I think is fine, we need to support it anyway.

Yes. That's the reason why I am choosing it,



> I wonder if it is time to promote the array descriptor into the type system (and also promote Array to become its own `ttype`), since that seems to be a clean way to do this: #1601.

I would do it in the following steps,

1. First support https://github.com/lfortran/lfortran/pull/1630#issuecomment-1536231630.
2. Then use `Array` as a `ttype`.
3. Then support wait and see if there are any bug which point to design flaws in the implementations added in the above two steps. If not then promote array descriptor to ASR. I would keep it as a physical type.

I guess we can attempt to do steps 1 and 2 in parallel because 1 needs updates in the way we index things while 2 is just a representational (a good one) change in ASR so both should be independent of each other.

**Comment by czgdp1807** [comment]
Also I noticed another problem with our algorithm of `sum` intrinsic. It traverses in row major order but the array is indexed using column major order in the backend. So I guess we should reverse the indexing of array in `sum` intrinsic so that its cache optimal.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -246,11 +246,20 @@ jobs:
             examples/example_lmdif1
             examples/example_primes
             ctest
+      - name: Test Minpack (Fortran-Lang)
+        shell: bash -x -l {0}
+        run: |
+            git clone https://github.com/czgdp1807/minpack modern_minpack_01
+            cd modern_minpack_01
+            git fetch origin skip-ic-21
+            git checkout skip-ic-21
```
Comment (certik) [inline .github/workflows/CI.yml:255]:
```suggestion
            git checkout skip-ic-21
            git checkout 765476455
```

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
I think this is good enough to merge after polishing the history.

**Pull request merged**